### ISSUE TITLE
Separate repository of rules for snaffling

### ISSUE THREAD

**Issue by ss23** [SnaffCon/Snaffler#78]
It would be great if the repository of what to snaffle was kept separate to the Snaffler tool itself. This allows people to build alternative implementations, and most importantly, update the two repositories independently (which makes it a lot easier to contribute to rules).

**Comment by hkelley** [comment]
I like this concept.  How would the repo be structured?  I wonder if there could be something even "lighter" than a full repo,  maybe just gists with rule "chunks".   

For example,  I have a set of Cisco-finding chunks I've been working on.   I wouldn't want to build a whole .toml file in a repo just for this.

```
# Cisco config rules
[[ClassifierRules]]
EnumerationScope = "FileEnumeration"
RuleName = "CiscoContentByName"
MatchAction = "Relay"
RelayTargets = ["KeepCiscoRegexRed","KeepCiscoRegexYellow"]
Description = "Files with these extensions will be searched for Cisco bits."
MatchLocation = "FileName"
WordListType = "Contains"
MatchLength = 0
WordList = ["conf"]
Triage = "Green"

[[ClassifierRules]]
EnumerationScope = "FileEnumeration"
RuleName = "CiscoContentByExt"
MatchAction = "Relay"
RelayTargets = ["KeepCiscoRegexRed","KeepCiscoRegexYellow"]
Description = "Files with these extensions will be searched for Cisco bits."
MatchLocation = "FileExtension"
WordListType = "Exact"
MatchLength = 0
WordList = [".txt"]
Triage = "Green"


[[ClassifierRules]]
EnumerationScope = "ContentsEnumeration"
RuleName = "KeepCiscoRegexRed"
MatchAction = "Snaffle"
Description = "Looking for Cisco stuff."
MatchLocation = "FileContentAsString"
WordListType = "Regex"
MatchLength = 0
WordList = ["enable secret\\s[0-7]","enable\\spassword","password 7","pac key [0-7] ","snmp-server community\\s.+\\sRW"]
Triage = "Red"

[[ClassifierRules]]
EnumerationScope = "ContentsEnumeration"
RuleName = "KeepCiscoRegexYellow"
MatchAction = "Snaffle"
Description = "Looking for Cisco stuff."
MatchLocation = "FileContentAsString"
WordListType = "Regex"
MatchLength = 0
WordList = ["enable secret [8-9]", "snmp-server community\\s.+\\sRO"]
Triage = "Yellow"

```

**Comment by l0ss** [comment]
If we were to do this it would probably make sense to have it as a repo containing a directory structure full of toml files containing rules grouped by some kind of coherent taxonomy.

Then we could maybe set up Snaffler's build process to bake those files into the exe as resources and have it deserialise and ingest them at runtime, just like it does at the moment with a toml config file.

What do you reckon?

**Comment by ss23** [comment]
That sounds great for what we were hoping to use the rules for! 👍

**Comment by l0ss** [comment]
OK I have started in on this, take a look and let me know what you think before I go too much further with it:

On the left you can see part of the overall structure for the 'look inside code files' section.

You can see that for each language we have a single "relay" rule that takes files that match the appropriate extension and relays them on to multiple other rules. Some of these rules are specific to the language (see the blue and pink arrows below for Python and PHP respectively) , so the files live in that dir, but some are more generic (like the one that looks for AWS keys) so they live further up in the directory structure (see the green arrow). This way if we make a change to refine the regex for something that is common across multiple file types, it should only need to be changed once in a single rule file.

tl;dr it's way more modular now.

<img width="1308" alt="image" src="https://user-images.githubusercontent.com/24580473/152678671-647c28e8-9d78-4af2-a5b2-8b5cb4765e3b.png">

Take a look in newrules branch, I've got a default.toml in the root of the RulesRepo dir that I've been deleting lines from as I replace them with stuff under the new structure, so I can keep track of what has made it in and what hasn't.

Let me know what you think.

**Comment by hkelley** [comment]
They layout in https://github.com/SnaffCon/Snaffler/tree/newrules/RuleRepo folder and image above look good.

I haven't yet dug through the code enough to understand how the .toml file rules and the "resource" rules will be blended.   Do you still expect to have a way to deliver custom rules via TOML?   Will there be a way to update "core" rules (particularly for domain users) at runtime?   

I ask because our domain user rule list becomes too expensive/slow if I take them all from AD, so I only hunt for higher-value identities.  I feed this list (as a regex) into our .toml file before I start scanning  (hooking on the "stub" below.    Similarly,  I might want to hunt for a particular server name, common password, etc.

```
[[ClassifierRules]]
EnumerationScope = "ContentsEnumeration"
RuleName = "KeepUsernameRegexYellow"
MatchAction = "Snaffle"
Description = "Matches on potential service account names."
MatchLocation = "FileContentAsString"
WordListType = "Regex"
MatchLength = 0
WordList = [USERNAME_REGEX_STUB]
Triage = "Yellow"
```

**Comment by l0ss** [comment]
I haven’t made any of the code changes yet but yes there will be a way to programmatically change them at runtime.

**Comment by l0ss** [comment]
OK I'm more awake now so here's roughly how I expect it to go, in busted pseudocode:

```
if (user has passed a toml config file using an arg)
{
    Deserialise toml to config
    return config
}
else if (user has submitted a directory full of toml rules using an arg)
{
    create a new config with defaults (which will no longer include default rules)
    override any default values using the user-provided args
    find all the .toml rule files in the user-defined dir and cat them into one big toml doc
    deserialise that toml doc into ruleset, stick it in config
    return config
}
else
{
    create a new config with defaults
    override any default values using the user-provided args
    Cat all the "default" .toml rule files stored as resources into one big toml doc
    deserialise toml doc into ruleset, stick it in config
    return config
}
```

The simplest approach would probably be to do some kind of find/replace in the toml-formatted ruleset before it gets deserialised, what do you think?

**Comment by l0ss** [comment]
The process of going through the rules like this has flushed out a bunch of duplication and gaps. 
It's good.

**Comment by hkelley** [comment]
I'm totally flexible about the "overwrite" mechanism for rules.   I was envisioning TOML-based hashtable where the rule name is the key and the wordlist is the value.    That could be fixed-up after all of the rules have been loaded.

I've never made such a beast in TOML, but this thread gives me hope it can be done:

[parsing-a-key-value-pair-within-a-table-for-toml-file-and-golang](https://stackoverflow.com/questions/53915596/parsing-a-key-value-pair-within-a-table-for-toml-file-and-golangl)

**Comment by l0ss** [comment]
Actually now we can have multiple relay targets per rule - we can just: 

- bake in a single "KeepUsernameInContent" rule with an empty wordlist
- if username searching is turned on, shove the resulting username regexen into that one rule
- have an easily-edited list of suitable Relay rules and add our KeepUsernameInContent rule to the RelayTargets array of each
- if username searching is turned off then that rule has an empty wordlist and is not a relay target so it's not a problem.

**Comment by l0ss** [comment]
OK first pass of making the thing work with a big stack of toml files is in at https://github.com/SnaffCon/Snaffler/tree/newrules

Default rules (which I haven't quite finished building from the old ones) are under Snaffler/DefaultRules - please take a look and let me know what you reckon.

**Comment by hkelley** [comment]
Folder layout looks good.  I nosed around in Config.cs but I'm not sure I see how to point it at a folder root for loading the files.

**Comment by l0ss** [comment]
Yeah I haven't added that part yet but I don't anticipate it will be hard.

**Comment by l0ss** [comment]
@ss23 does having a directory full of toml files (rather than C# code) in the existing Snaffler repo meet your needs for this issue or is there some specific reason to have a separate repository?

**Comment by ss23** [comment]
Tagging @SoftPoison who can answer the question for me :)

**Comment by SoftPoison** [comment]
Hey @l0ss, thanks for your work on this! The changes look great :)

I think it would be better if it were to be a separate repo, so that alternative implementations can just submodule in the rules without having to include the entire repo with the C# code. It would also allow for the rules to be forked and changed without needing to fork the code too.

**Comment by l0ss** [comment]
I'm OK with doing this, but it would be a super helpful timesaver if someone who is better at both git and VS could show me how to set up a submodule on the existing repo, and how to set up a build task to programmatically add every file in the dir as an embedded resource for the Snaffler project.

**Comment by SoftPoison** [comment]
Adding the rules as a git submodule should be as easy as `git submodule add https://github.com/SnaffCon/SnafflerRules.git ./rules; git commit; git push`.

From doing some quick research (and from limited experience), I think this is how embedding the rules files in the exe works:

Embed a folder of files in C#:
- https://stackoverflow.com/a/70846148

Get embedded files from assembly pseudo-C#:
```c#
var assembly = Assembly.GetExecutingAssembly();
var resourceNames = assembly.GetManifestResourceNames();
// filter the names so that only rules files are included (probably by checking if they end in .toml, or the name contains .rules. or something)

// for name in resourceNames
Stream stream = assembly.GetManifestResourceStream(name);
StreamReader reader = new StreamReader(stream);
var fileContents = reader.ReadToEnd();
// Close the reader and stream
```

I may be wrong though, so someone with more C# knowledge please correct me!

**Comment by l0ss** [comment]
Thanks, I think that did the trick.

I've implemented that in the newrules branch, and the new repo under SnaffCon/SnaffRules.

Please test and let me know if it's all still working as expected. If so, I'll implement HK's request to have snaffler consume a dir full of toml rules at runtime, and we can finish cleaning up the new rule repo.

**Comment by l0ss** [comment]
@hkelley ok `-p C:\temp\ruledir` will now load every .toml file in that directory (recursively) in place of the default ruleset.

**Comment by hkelley** [comment]
Is the vision that we supply both -z and -p from the command line (ignoring any rules that may be in the -z file),  merge the two rule sets, or other?

**Comment by l0ss** [comment]
Ah, i believe you've uncovered a logic bug in the new rules setup, where the config file parsing and the rule dir stuff was happening in the wrong order. I'll see if i can iron that out tonight then explain what i had in mind.

**Comment by l0ss** [comment]
btw i might just be bad at git, but this submodule/separate-repo thing is making it a pain in the arse to manage.

**Comment by l0ss** [comment]
OK @hkelley here's how it should work now:

as before, if you tack `-z generate` onto the end of a command line it will make you a config file with all the settings defined in the rest of the args.

so for example, if you do `snaffler.exe -s -i D:\ -v degub -p C:\hk-rules -z generate` it will make a config file that will make snaffler load the rules out of C:\hk-rules every time you run it.

if a config file has rules inside it, those will override all other rules.

if a config file has no rules in it and you don't pass the -p arg, the default ruleset gets loaded from resources embedded in the exe.

what do you think?

**Comment by l0ss** [comment]
I believe I've fixed up #88, more detail in that issue.
